### PR TITLE
Current Logic User CC Info

### LINKED ISSUE #104: Get Current User CC Info
- Endpoint URI: */user-cc-infos/current*
- Http Method: *GET*

- API Logic: 
    - Get current login user all cc info

- Progress
    - [x] development
    - [x] integration Test

### PR DESCRIPTION
**Pull request by HowardYaw**
resolved #104

### CODE REVIEW

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java [hunk 1]
```diff
@@ -91,6 +99,83 @@ public UserCCInfoDTO save(UserCCInfoDTO userCCInfoDTO) {
...
+                }
+                if (currentYearSession.compareTo(secondYearSession) > 0) {
+                    String thirdYearSession = YearSessionUtils.addYearSessionWithSemester(secondYearSession, 2) + " - Present";
+                    fullUserCCInfoDTOList.add(constructCustomUserCCInfo(thirdYearSession, FishLevel.ELDER_FISH, userId));
+                }
+                userCCInfoDTOList.stream()
+                    .filter(userCCInfoDTO1 -> userCCInfoDTO1.getYearSession().compareTo(secondYearSession) > 0)
+                    .forEach(fullUserCCInfoDTOList::add);
+            });
+        fullUserCCInfoDTOList.sort(Comparator.comparing(UserCCInfoDTO::getYearSession));
```
Comment (kachun333) [inline src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: This code is too hard to read and maintain.
Suggest you use a loop.
- increment yearSession by 1 every time
- determine fish level based on loop index `i`
- then no need to sort using sort function also, since it will guarantee to be in ascending Year Session

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java [hunk 2]
```diff
@@ -91,6 +99,83 @@ public UserCCInfoDTO save(UserCCInfoDTO userCCInfoDTO) {
...
+        return fullUserCCInfoDTOList;
+    }
+
+    private UserCCInfoDTO constructCustomUserCCInfo(String yearSession, FishLevel fishLevel, Long userId) {
```
Comment (kachun333) [inline src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: `constructCustomUserCCInfo` is quite a weird name.
probably `mapUserCCInfoDTO` is a better name.

or perhaps you might not even need this private function if you can refractor to use loop

**Review by kachun333 — DISMISSED** [review]
The test scripts are very well written. 

Unfortunately, there is a part of the code that is incredibly problematic and hard to maintain. 
I have dropped some suggestions and requesting a change. If you need help, feel free to reach out.

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java [hunk 3]
```diff
@@ -91,6 +99,83 @@ public UserCCInfoDTO save(UserCCInfoDTO userCCInfoDTO) {
...
+     * @param userId
+     * @return List of user CC Info Sorted by Year Session
+     */
+    private List<UserCCInfoDTO> getFullUserCCInfoList(List<UserCCInfoDTO> userCCInfoDTOList, Long userId) {
+        List<String> yearSessionWithCCFamilyRole = userCCInfoDTOList.stream()
+            .map(UserCCInfoDTO::getYearSession)
+            .collect(Collectors.toList());
+        List<UserCCInfoDTO> fullUserCCInfoDTOList = new ArrayList<>();
+        userUniInfoService.getUserUniInfoByUserId(userId)
+            .ifPresent(userUniInfoDTO -> {
```
Comment (kachun333) [inline src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: what if `Optional optional = userUniInfoService.getUserUniInfoByUserId(userId)` is `Optional.empty()`?
what will happen?

lengthy lambad function is hard to read. I prefer using 
```java
if (!optional.isPresent()) {
// not found 
 return smtg;
}
Object obj = optional.get();
// do further processing if found
return obj;
```

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java [hunk 4]
```diff
@@ -91,6 +99,83 @@ public UserCCInfoDTO save(UserCCInfoDTO userCCInfoDTO) {
...
+     * First Year Session is "Junior Fish"
+     * Second Year Session is "Senior Fish"
+     * Later Year Session until Present is "Elder Fish"
+     *
+     * @param userCCInfoDTOList
+     * @param userId
+     * @return List of user CC Info Sorted by Year Session
+     */
+    private List<UserCCInfoDTO> getFullUserCCInfoList(List<UserCCInfoDTO> userCCInfoDTOList, Long userId) {
+        List<String> yearSessionWithCCFamilyRole = userCCInfoDTOList.stream()
```
Comment (kachun333) [inline src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: `yearSessionWithCCFamilyRole` is an inappropriate name. Since the list only contains yearsession string

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java [hunk 5]
```diff
@@ -91,6 +99,83 @@ public UserCCInfoDTO save(UserCCInfoDTO userCCInfoDTO) {
...
+                        UserCCInfoDTO userCCInfoDTO = userCCInfoDTOList.stream()
+                            .filter(userCCInfoDTO1 -> userCCInfoDTO1.getYearSession().equals(secondYearSession))
+                            .findFirst()
+                            .get();
+                        userCCInfoDTO.setFishLevel(FishLevel.SENIOR_FISH);
+                        fullUserCCInfoDTOList.add(userCCInfoDTO);
+                    }
+                }
+                if (currentYearSession.compareTo(secondYearSession) > 0) {
+                    String thirdYearSession = YearSessionUtils.addYearSessionWithSemester(secondYearSession, 2) + " - Present";
```
Comment (kachun333) [inline src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: adding " - Present" is not clean leh.

why is this needed ya?
just return normal "2020/2021" can?

if you format the string data in this weird way, what if other modules need to consume this data?

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java (hunk 1 above)
Comment (HowardYaw) [reply src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: @kachun333 
Can I separate the if into 3 diff methods instead of using loop?

What I think is Loop performance is slower

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java (hunk 4 above)
Comment (HowardYaw) [reply src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: @kachun333 
this is the list of year session that user have CC Family Role
maybe I name it `yearSessionListHaveCCFamilyRole` ?

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java (hunk 5 above)
Comment (HowardYaw) [reply src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: ok will do the change

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java (hunk 3 above)
Comment (HowardYaw) [reply src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: actually if Optional.empty() 
no need to do anything

Anyway I will change since there is a request

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java (hunk 5 above)
Comment (HowardYaw) [reply src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: resolved in d5253a35de87edf18588bc0c2bae3356b925f969

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java (hunk 4 above)
Comment (HowardYaw) [reply src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: resolved in d5253a35de87edf18588bc0c2bae3356b925f969

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java (hunk 3 above)
Comment (HowardYaw) [reply src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: resolved in d5253a35de87edf18588bc0c2bae3356b925f969

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java (hunk 2 above)
Comment (HowardYaw) [reply src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: resolved in d5253a35de87edf18588bc0c2bae3356b925f969

File: src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java (hunk 1 above)
Comment (HowardYaw) [reply src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java]: resolved in d5253a35de87edf18588bc0c2bae3356b925f969

**Review by kachun333 — DISMISSED** [review]

File: src/test/java/com/thirdcc/webapp/utils/YearSessionUtilsTest.java [hunk 6]
```diff
@@ -101,14 +102,42 @@ public void getFirstInstantOfYearSession() {
...
+            YearSessionUtils.isBefore(validYearSession, invalidYearSession);
+        });
+        assertThat(thrown2).isInstanceOf(IllegalArgumentException.class);
+    }
 
-        assertThat(actualMessage).isEqualTo(expectedMessage);
+    @Test
+    public void isBefore_WithTrue() {
+        String earlierYearSession = "2019/2020";
+        String laterYearSession = "2020/2021";
```
Comment (LUXIANZE) [inline src/test/java/com/thirdcc/webapp/utils/YearSessionUtilsTest.java]: make test class level constant

File: src/test/java/com/thirdcc/webapp/web/rest/UserCCInfoResourceIT.java [hunk 7]
```diff
@@ -21,8 +31,9 @@
 import javax.persistence.EntityManager;
 import java.util.List;
 
+import static com.thirdcc.webapp.utils.FishLevelUtils.*;
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.hamcrest.Matchers.hasItem;
+import static org.hamcrest.Matchers.*;
```
Comment (LUXIANZE) [inline src/test/java/com/thirdcc/webapp/web/rest/UserCCInfoResourceIT.java]: should avoid using wildcard import

Comment (LUXIANZE) [inline src/test/java/com/thirdcc/webapp/web/rest/UserCCInfoResourceIT.java]: should avoid using wildcard import

File: src/test/java/com/thirdcc/webapp/utils/YearSessionUtilsTest.java (hunk 6 above)
Comment (LUXIANZE) [inline src/test/java/com/thirdcc/webapp/utils/YearSessionUtilsTest.java]: make test class level constant

**Review by LUXIANZE — CHANGES_REQUESTED** [review]

File: src/test/java/com/thirdcc/webapp/web/rest/UserCCInfoResourceIT.java (hunk 7 above)
Comment (HowardYaw) [reply src/test/java/com/thirdcc/webapp/web/rest/UserCCInfoResourceIT.java]: resolved in bdd2e2dae928ebadc72909909ce8fe17c5081841

File: src/test/java/com/thirdcc/webapp/utils/YearSessionUtilsTest.java (hunk 6 above)
Comment (HowardYaw) [reply src/test/java/com/thirdcc/webapp/utils/YearSessionUtilsTest.java]: resolved in 2d7fc372151eeb81ae71d52f085f4d44d70af84b

**Review by LUXIANZE — APPROVED** [review]

**Review by CarlosSia — APPROVED** [review]
